calcom/cal.com
# feat: Round Robin Weights

## What does this PR do?

Adds weights to round robin alogirthm. When weights are enabled each host has 100% as the default weight. That means that ideally, all hosts end up with the same amount of bookings. If weights are enabled, priorities are only taken into account when more than one host is selected by the `getUsersBasedOnWeights` function.

To make sure that adding new host, when other hosts already have bookings doesn't result in the new host being assigned to all bookings we use weightAdjustments. 

`WeightAdjustment (nr of 'fake' bookings)` = ((`nr of all bookings of current RR hosts` + `sum of all weight adjustments of current RR hosts`) / `sum of all current RR hosts weights`) * `new host's weight`

Weights enabled: 
<img width="999" alt="Screenshot 2024-06-25 at 4 30 13 PM" src="https://github.com/calcom/cal.com/assets/30310907/03559f97-ea66-4a31-b613-be1ee33f12bb">

Weights disabled: 
<img width="994" alt="Screenshot 2024-06-25 at 4 30 29 PM" src="https://github.com/calcom/cal.com/assets/30310907/ad9f4fa5-ba12-4715-bb9d-fa6f14f5bb44">

Set weight: 
<img width="749" alt="Screenshot 2024-06-25 at 4 31 00 PM" src="https://github.com/calcom/cal.com/assets/30310907/c30bcf62-338a-45ad-b2e1-1d9a94ba8bb2">

- Fixes #15340
- Fixes CAL-3869

## Mandatory Tasks (DO NOT REMOVE)

- [ ] I have added a Docs https://github.com/calcom/docs/pull/101 --> still need to finish that

## How should this be tested?

- Create RR event type and enable weights
- test different weights and priority settings 
- test adding new hosts when bookings already exists and see if weightAdjustment was correctly saved
